lib/src/front_end/ast_node_visitor.dart

```diff
@@ -1214,104 +1254,97 @@ class AstNodeVisitor extends ThrowingAstVisitor<void>
   }
 
   @override
-  void visitShowCombinator(ShowCombinator node) {
-    assert(false, 'Combinators are handled by createImport().');
+  Piece visitShowCombinator(ShowCombinator node) {
+    throw UnsupportedError('Combinators are handled by createImport().');
   }
 
   @override
-  void visitSimpleFormalParameter(SimpleFormalParameter node) {
-    startFormalParameter(node);
-
-    if ((node.type, node.name) case (var type?, var name?)) {
-      // Have both a type and name, so allow splitting between them.
-      modifier(node.keyword);
-      visit(type);
-      var typePiece = pieces.split();
+  Piece visitSimpleFormalParameter(SimpleFormalParameter node) {
+    var builder = AdjacentBuilder(this);
+    startFormalParameter(node, builder);
+    builder.modifier(node.keyword);
+    builder.visit(node.type);
 
-      token(name);
-      var namePiece = pieces.take();
-
-      pieces.give(VariablePiece(typePiece, [namePiece], hasType: true));
+    if ((node.type, node.name) case (var _?, var name?)) {
```

Comment:
[optional] This would be nicer if the fields were destructured by the pattern instead of manually into a record, but the long type name would be ugly. Makes me hope more for https://github.com/dart-lang/language/issues/2563 so this could be something like `if (node case #(type: TypeAnnotation, :var name?))`. I'm 50/50 on whether it's worth doing with the full type name.

```suggestion
    if (node case SimpleFormalParameter(type: TypeAnnotation, :var name?)) {
```

Reply:
Yeah, inferring the object pattern name would help a lot here. I think the current code is OK.